cfgrammar/src/lib/yacc/parser.rs

```diff
@@ -310,20 +310,29 @@ impl YaccParser {
     }
 
     fn parse_rules(&mut self, mut i: usize) -> YaccResult<usize> {
+        let mut duplicate_spans = Vec::new();
         // self.parse_declarations should have left the input at '%%'
         i = self.lookahead_is("%%", i).unwrap();
         i = self.parse_ws(i, true)?;
         while i < self.src.len() {
             if self.lookahead_is("%%", i).is_some() {
                 break;
             }
-            i = self.parse_rule(i)?;
+            i = self.parse_rule(&mut duplicate_spans, i)?;
             i = self.parse_ws(i, true)?;
         }
+
+        if let Some(span) = duplicate_spans.first() {
+            return Err(YaccParserError {
+                kind: YaccParserErrorKind::DuplicateRule,
+                span: *span,
+            });
+        }
+
         Ok(i)
     }
 
-    fn parse_rule(&mut self, mut i: usize) -> YaccResult<usize> {
+    fn parse_rule(&mut self, duplicate_spans: &mut Vec<Span>, mut i: usize) -> YaccResult<usize> {
```

Comment:
This seems like it may want to be a `HashMap<Name, Vec<Span>>`, if we want to collate these by name.
rather than the order they appear in?

Reply:
Good point -- I agree.